lib/core/include/scipp/core/dict.h

```diff
@@ -0,0 +1,423 @@
+
+public:
+  using difference_type = std::ptrdiff_t;
+  using value_type = typename ValueType<It1, It2>::type;
+  using pointer = std::add_pointer_t<std::remove_reference_t<value_type>>;
+  using reference = typename ReferenceType<It1, It2>::type;
+
+  template <class T>
+  Iterator(std::reference_wrapper<Container> container, T &&it1)
+      : m_iterators{std::forward<T>(it1)}, m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  template <class T, class U>
+  Iterator(std::reference_wrapper<Container> container, T &&it1, U &&it2)
+      : m_iterators{std::forward<T>(it1), std::forward<U>(it2)},
+        m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  decltype(auto) operator*() const {
+    expect_container_unchanged();
+    if constexpr (sizeof...(IteratorIndices) == 1) {
+      return *std::get<0>(m_iterators);
+    } else {
+      return std::make_pair(std::cref(*std::get<0>(m_iterators)),
+                            std::ref(*std::get<1>(m_iterators)));
+    }
+  }
+
+  decltype(auto) operator->() const {
+    if constexpr (sizeof...(IteratorIndices) == 1) {
+      expect_container_unchanged();
+      return std::get<0>(m_iterators);
+    } else {
+      return TemporaryItem<reference>(**this);
+    }
+  }
+
+  Iterator &operator++() {
+    expect_container_unchanged();
+    (++std::get<IteratorIndices>(m_iterators), ...);
+    return *this;
+  }
+
+  bool operator==(
+      const Iterator<Container, It1, It2, IteratorIndices...> &other) const {
+    expect_container_unchanged();
+    // Assuming m_iterators are always in sync.
+    return std::get<0>(m_iterators) == std::get<0>(other.m_iterators);
+  }
+
+  bool operator!=(
+      const Iterator<Container, It1, It2, IteratorIndices...> &other) const {
+    return !(*this == other);
+  }
+
+  template <class F> auto transform(F &&func) const & {
+    return TransformIterator{*this, std::forward<F>(func)};
+  }
+
+  template <class F> auto transform(F &&func) && {
+    return TransformIterator{std::move(*this), std::forward<F>(func)};
+  }
+
+  friend void swap(Iterator &a, Iterator &b) {
+    swap(a.m_iterators, b.m_iterators);
+    swap(a.m_container, b.m_container);
+    std::swap(a.m_end_address, b.m_end_address);
+  }
+
+protected:
+  // operator-> needs to return a pointer or something that has operator->
+  // But we cannot take the address of the temporary pair or transform result.
+  // So store it in this wrapper to make it accessible via its address.
+  template <class T> class TemporaryItem {
+  public:
+    explicit TemporaryItem(T &&item) : m_item(std::move(item)) {}
+    auto *operator->() { return &m_item; }
+
+  private:
+    std::decay_t<T> m_item;
+  };
+
+private:
+  using IteratorStorage =
+      std::tuple<typename std::tuple_element<IteratorIndices,
+                                             std::tuple<It1, It2>>::type...>;
```

Comment:
The reason was that I wanted to avoid doing something like
```.cpp
++std::get<0>(m_iterators);
if constexpr(sizeof...(It) == 2)
  ++std::get<1>(m_iterators);
```
in every function. But looking at it now, `operator++` is the only one that actually uses `IteratorIndices`. So I can just remove the indices and use the code shown above.